staging/src/k8s.io/client-go/tools/portforward2/portforward2.go

```diff
@@ -0,0 +1,314 @@
+	errWriter         io.Writer
+	requestID         uint32
+	activeConnections sync.WaitGroup
+}
+
+// New creates a new PortForwarder2 for localhost.
+func New(dialer httpstream.Dialer, ports []string, outWriter io.Writer, errWriter io.Writer) *PortForwarder2 {
+	return NewOnAddresses(dialer, []string{"localhost"}, ports, outWriter, errWriter)
+}
+
+// NewOnAddresses creates a new PortForwarder2 with custom listen addresses.
+func NewOnAddresses(dialer httpstream.Dialer, addresses []string, ports []string, outWriter io.Writer, errWriter io.Writer) *PortForwarder2 {
+	return &PortForwarder2{
+		dialer:    dialer,
+		addresses: addresses,
+		ports:     ports,
+		outWriter: outWriter,
+		errWriter: errWriter,
+	}
+}
+
+// ForwardPorts starts port forwarding and continues until the provided context is cancelled.
+func (pf *PortForwarder2) ForwardPorts(ctx context.Context) error {
+	// Keep track of listeners, so they can be closed later
+	var listeners []net.Listener
+
+	// Create a new context, with cancel, so that we can cancel the listeners to stop gracefully
+	listenerContext, cancelListeners := context.WithCancel(ctx)
+
+	// Stop all listeners and wait for all active connections to close
+	defer func() {
+		cancelListeners()
+		for _, listener := range listeners {
+			if err := listener.Close(); err != nil {
+				pf.errorf("failed to close listener: %s", err) // Print the error, but don't fail during cleanup
+			}
+		}
+		pf.activeConnections.Wait()
+	}()
+
+	// Create a listener for each listenAddress/port combination
+	for _, a := range pf.addresses {
+		for _, p := range pf.ports {
+			listener, err := pf.listen(listenerContext, a, p)
+			if err != nil {
+				return err
+			}
+			listeners = append(listeners, listener)
+		}
+	}
+
+	// Wait for context to be done
+	<-listenerContext.Done()
+
+	return nil
+}
+
+func (pf *PortForwarder2) listen(ctx context.Context, address string, port string) (net.Listener, error) {
+	protocol, err := determineAddressProtocol(address)
+	if err != nil {
+		return nil, err
+	}
+
+	localPort, remotePort, err := parseLocalAndRemotePorts(port)
+	if err != nil {
+		return nil, err
+	}
+
+	addressWithPort := net.JoinHostPort(address, localPort)
+	listener, err := net.Listen(protocol, addressWithPort)
+	if err != nil {
+		return nil, fmt.Errorf("listen failed on %s: %s", addressWithPort, err)
+	}
+
+	listenerAddress := listener.Addr().String()
+	pf.printf("Forwarding from %s -> %s\n", listenerAddress, remotePort)
+
+	// Accept and handle connections
+	go func() {
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					pf.errorf("accept failed: %s\n", err)
+					continue
+				}
+			}
+			go pf.handleConnection(ctx, conn, listenerAddress, remotePort)
+		}
+	}()
+
+	return listener, nil
+}
+
+func (pf *PortForwarder2) handleConnection(ctx context.Context, localConn net.Conn, listenerAddress string, remotePort string) {
+	defer localConn.Close()
```

Comment:
if you exist this method you close the local connection, how do you solve the problem?
My understanding is that this creates a local proxy like

client --> [listerner](kubectl)[conntoapiserver] --->{magic}-->Pod

if kubectl to pod breaks the stream, do you plan to resume that stream?

Reply:
You are correct about the local connection being closed when handleConnection exits.

This should be fine though because it blocks on `doneChan` or if the context is cancelled (triggered by SIGTERM for example):

```
	select {
	case <-ctx.Done():
		remoteDataStream.Reset() // Reset data stream to unblock io.Copy
	case <-doneChan:
	}
```

doneChan is closed after the `io.Copy()` to read from the remote stream has completed.  So this would be one of these two situations:
1. Remote connection returned EOF - the normal case
2. Remote connection was lost and `io.Copy()` returns an error

Either way, this particular connection is finished at this point.  PortForwarder is still listening and able to attempt another connection though, which is the improvement.  Currently, PortForwarder is "ruined", for lack of a better term, when this happens and has to be completely restarted.

For what its worth, the handleConnection implementation is pretty similar to how the existing PortForwarder works.  The main difference is that handleConnection dials to establish a new remote connection each time instead of trying to reuse a single remote connection forever.

Reply:
but there is no way to resume a TCP connection ... I may be not understanding well the problem, but looks to me that redialing just create a new connection from scratch